### PR TITLE
Consolidate IR: MatrixRead

### PR DESCRIPTION
**Pull request by jigold**
@cseed @tpoterba Is this what you had in mind? I'm guessing the functions I moved from the IR nodes to MatrixTable will have to end up somewhere else eventually.

### CODE REVIEW

File: src/main/scala/is/hail/expr/Relational.scala [hunk 1]
```diff
@@ -407,13 +407,11 @@ case class MatrixLiteral(
...
-  spec: MatrixTableSpec,
+  typ: MatrixType,
+  override val partitionCounts: Option[Array[Long]],
   dropCols: Boolean,
-  dropRows: Boolean) extends MatrixIR {
-  def typ: MatrixType = spec.matrix_type
-
-  override def partitionCounts: Option[Array[Long]] = Some(spec.partitionCounts)
+  dropRows: Boolean,
+  f: (HailContext) => MatrixValue) extends MatrixIR {
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Relational.scala]: I think this f should be MatrixRead => MatrixValue

**Comment by tpoterba** [comment]
I'll review, but we should wait until Dan's meatball goes in because this would definitely conflict.

File: src/main/scala/is/hail/expr/Relational.scala (hunk 1 above)
Comment (jigold) [reply src/main/scala/is/hail/expr/Relational.scala]: I don't see why you'd want the input to be MatrixRead. Should I close this for now and we can discuss next week?

Comment (tpoterba) [reply src/main/scala/is/hail/expr/Relational.scala]: sure, but basically we need to be able to push down dropRows here

Comment (jigold) [reply src/main/scala/is/hail/expr/Relational.scala]: ok. that makes sense.

File: src/main/scala/is/hail/variant/MatrixTable.scala [hunk 2]
```diff
@@ -126,8 +126,89 @@ object MatrixTable {
...
     }
+
+    val typ = spec.matrix_type
+
+    val f: (MatrixRead) => MatrixValue = { mr =>
+      val hc = HailContext.get
+      val globals = Annotation.copy(typ.globalType, spec.globalsComponent.readLocal(hc, path)(0)).asInstanceOf[Row]
+
+      val colAnnotations =
+        if (dropCols)
```
[the remark this answers is not in the document]
Comment (jigold) [reply src/main/scala/is/hail/variant/MatrixTable.scala]: Where does the user telling Hail to drop rows or cols go then?

**Comment by tpoterba** [comment]
Ah, right. Eventually we can remove those when we can optimize them away, but I guess we need them now

**Comment by jigold** [comment]
Do you have suggestions for the `Pretty` code? I think having the path is important -- maybe it should still be a class parameter?  I also thought @cseed wanted two different types. I can't remember what the second one was.

**Comment by tpoterba** [comment]
I don't much care. I generally think more information is better, but think the path is probably one of the least useful pieces of info (if we're debugging IR, the file path is probably on hand anyway).

> wanted two different types
can you elaborate?

**Comment by jigold** [comment]
I thought it was the requested type and the type that will actually be computed, but I can't remember.

**Comment by tpoterba** [comment]
oh, the file type and the requested type. It seems like the file type can be hidden inside of the function though -- I'm not sure why we need it in the node.

**Pull request merged**